WordPress/gutenberg
# [Inserter]: Fix reset of registered media categories

## What?
Fixes: https://github.com/WordPress/gutenberg/issues/52832

In https://github.com/WordPress/gutenberg/pull/51542 a new `registerInserterMediaCategory` API to make media categories extensible was introduced. Since `inserterMediaCategories` is a block editor setting though there was an issue that if we register new categories before the initialization of the block editor store's settings or we updated another setting that triggers a settings reset(ex spotlight mode), the added categories were lost.

This PR introduces a new reducer in block editor(`registeredInserterMediaCategories`) and merges the registered ones and the settings when consuming the media categories in media tab.



## Testing Instructions
1. Register categories with PHP and JS
2. Update a setting like `Spotlight mode`
3. Observe that the categories are preserved.

PHP registration code:
```php
add_action( 'enqueue_block_editor_assets', function() {
	wp_add_inline_script( 'wp-editor', "
wp.domReady( function() {
	wp.data.dispatch( 'core/block-editor' ).registerInserterMediaCategory( {
		name: 'openverse-2',
		labels: {
			name: 'Openverse 2',
			search_items: 'Search Openverse',
		},
		mediaType: 'image',
		async fetch( query = {} ) {
			const defaultArgs = {
				mature: false,
				excluded_source: 'flickr,inaturalist,wikimedia',
				license: 'pdm,cc0',
			};
			const finalQuery = { ...query, ...defaultArgs };
			const mapFromInserterMediaRequest = {
				per_page: 'page_size',
				search: 'q',
			};
			const url = new URL( 'https://api.openverse.engineering/v1/images/' );
			Object.entries( finalQuery ).forEach( ( [ key, value ] ) => {
				const queryKey = mapFromInserterMediaRequest[ key ] || key;
				url.searchParams.set( queryKey, value );
			} );
			const response = await window.fetch( url, {
				headers: {
					'User-Agent': 'WordPress/inserter-media-fetch',
				},
			} );
			const jsonResponse = await response.json();
			const results = jsonResponse.results;
			return results.map( ( result ) => ( {
				...result,
				sourceId: result.id,
				id: undefined,
				caption: result.caption,
				previewUrl: result.thumbnail,
			} ) );
		},
		getReportUrl: ( { sourceId } ) =>
			'https://wordpress.org/openverse/image/' + sourceId + '/report/',
		isExternalResource: true,
	} );
	console.log( 'Registered inserter media categories', wp.data.select( 'core/block-editor' ).getSettings().inserterMediaCategories );
} );
	" );
} );
```

JS code to run in dev tools:
```
wp.data.dispatch( 'core/block-editor' ).registerInserterMediaCategory( {
	name: 'openverse from JS',
	labels: {
		name: 'Openverse JS',
		search_items: 'Search Openverse',
	},
	mediaType: 'image',
	async fetch( query = {} ) {
		const defaultArgs = {
			mature: false,
			excluded_source: 'flickr,inaturalist,wikimedia',
			license: 'pdm,cc0',
		};
		const finalQuery = { ...query, ...defaultArgs };
		// Sometimes you might need to map the supported request params according to `InserterMediaRequest`.
		// interface. In this example the `search` query param is named `q`.
		const mapFromInserterMediaRequest = {
			per_page: 'page_size',
			search: 'q',
		};
		const url = new URL( 'https://api.openverse.engineering/v1/images/' );
		Object.entries( finalQuery ).forEach( ( [ key, value ] ) => {
			const queryKey = mapFromInserterMediaRequest[ key ] || key;
			url.searchParams.set( queryKey, value );
		} );
		const response = await window.fetch( url, {
			headers: {
				'User-Agent': 'WordPress/inserter-media-fetch',
			},
		} );
		const jsonResponse = await response.json();
		const results = jsonResponse.results;
		return results.map( ( result ) => ( {
			...result,
			// If your response result includes an `id` prop that you want to access later, it should
			// be mapped to `InserterMediaItem`'s `sourceId` prop. This can be useful if you provide
			// a report URL getter.
			// Additionally you should always clear the `id` value of your response results because
			// it is used to identify WordPress media items.
			sourceId: result.id,
			id: undefined,
			caption: result.caption,
			previewUrl: result.thumbnail,
		} ) );
	},
	getReportUrl: ( { sourceId } ) =>
		`https://wordpress.org/openverse/image/${ sourceId }/report/`,
	isExternalResource: true,
} );
```
